core/src/main/java/org/apache/brooklyn/core/catalog/internal/BundleUpgradeParser.java

```diff
@@ -0,0 +1,239 @@
+            } else {
+                return new VersionRangedName(parts[0], "["+parts[1]+","+parts[1]+"]");
+            }
+        }
+
+        public VersionRangedName(String name, VersionRange v) {
+            this.name = checkNotNull(name, "name").toString();
+            this.v = checkNotNull(v, "versionRange").toString();
+        }
+        
+        private VersionRangedName(String name, String v) {
+            this.name = checkNotNull(name, "name");
+            this.v = checkNotNull(v, "versionRange");
+        }
+        
+        @Override
+        public String toString() {
+            return name + ":" + v;
+        }
+        
+        public String toOsgiString() {
+            return name + ":" + getOsgiVersionRange();
+        }
+
+        public String getSymbolicName() {
+            return name;
+        }
+
+        public VersionRange getOsgiVersionRange() {
+            if (cachedOsgiVersionRange == null) {
+                cachedOsgiVersionRange = VersionRange.valueOf(BrooklynVersionSyntax.toValidOsgiVersionRange(v));
+            }
+            return cachedOsgiVersionRange;
+        }
+
+        public String getVersionString() {
+            return v;
+        }
+        
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(name, v);
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (!(other instanceof VersionRangedName)) {
+                return false;
+            }
+            VersionRangedName o = (VersionRangedName) other;
+            return Objects.equal(name, o.name) && Objects.equal(v, o.v);
+        }
+    }
+
+    public static CatalogUpgrades parseBundleManifestForCatalogUpgrades(Bundle bundle) {
+        Dictionary<String, String> headers = bundle.getHeaders();
+        String removedLegacyItemsHeader = headers.get(MANIFEST_HEADER_FORCE_REMOVE_LEGACY_ITEMS);
+        String removedBundlesHeader = headers.get(MANIFEST_HEADER_FORCE_REMOVE_BUNDLES);
+        List<VersionRangedName> removedLegacyItems = ImmutableList.of();
+        List<VersionRangedName> removedBundles = ImmutableList.of();
+        if (removedLegacyItemsHeader == null && removedBundlesHeader == null) {
+            return CatalogUpgrades.EMPTY;
+        }
+        if (removedLegacyItemsHeader != null) {
+            removedLegacyItems = parseVersionRangedNameList(removedLegacyItemsHeader, false);
+        }
+        if (removedBundlesHeader != null) {
+            removedBundles = parseVersionRangedNameList(removedBundlesHeader, false);
+        }
+        return CatalogUpgrades.builder()
+                .removedLegacyItems(removedLegacyItems)
+                .removedBundles(removedBundles)
+                .build();
+    }
+    
+    @VisibleForTesting
+    static List<VersionRangedName> parseVersionRangedNameList(String input, boolean singleVersionIsOsgiRange) {
+        List<String> vals = QuotedStringTokenizer.builder()
```

Comment:
judging by the tests i'm surprised this works for mismatched brackets, or at least require a boolean flag to allow mismatched brackets -- but your test indicates otherwise

Reply:
It works if each `name:versionRange` is in quotes, with commas between them.

Reply:
sorry i confused the range test with the header test

need to document this well, ugly but not unreasonable.
i've added a good error message if user forgets them (also adds a trim if user puts spaces in):

https://github.com/apache/brooklyn-server/pull/870

i'll update that with any other tweaks i find too